Simplify the PMIx_Get processing:

rhc54 opened the pull request:
If the key starts with “pmix”, then it is a PMIx-namespace key and can only be in the data provided at startup (which is stored in the “internal" hash table). Otherwise, it is data that was PMix_Put and is in the “modex” hash table. PMIx job-level data is stored in the “internal” hash table against rank=PMIX_RANK_WILDCARD.

This reduces all calls to PMIx_Get to a single hash-table lookup

jladd-mlnx commented:
👍

rhc54 commented:
@dsolt Please see if this fixes the problem you observed

rhc54 commented:
I think the Jenkins server is picking up a stale aclocal.m4 that is leftover from the MCA pull request. I can't replicate a build problem with this PR, including doing ```make_dist_tarball```

artpol84 commented on src/client/pmix_client_get.c:
```diff
@@ -296,39 +296,46 @@ static void _getnb_cbfunc(struct pmix_peer_t *pr, pmix_usock_hdr_t *hdr,
-                                        "pmix: found requested value");
-                    if (PMIX_SUCCESS != (rc = pmix_bfrop.copy((void**)&val, cur_kval->value, PMIX_VALUE))) {
+                                        "pmix: unpacked key %s", cur_kval->key);
+                    if (PMIX_SUCCESS != (rc = pmix_hash_store(&nptr->modex, cur_rank, cur_kval))) {
```
This is Get, why do we `pmix_hash_store` here?

artpol84 replied:
Ok, I see. This is response and we store as well as return.

artpol84 commented on src/server/pmix_server_get.c:
```diff
@@ -404,6 +413,29 @@ static pmix_status_t _satisfy_request(pmix_nspace_t *nptr, int rank,
+     * from their own, then include a copy of the job-level info */
+    if (NULL != cd &&
+        0 != strncmp(nptr->nspace, cd->peer->info->nptr->nspace, PMIX_MAX_NSLEN)) {
+        cur_rank = PMIX_RANK_WILDCARD;
```
@rhc54 Does that mean that we will include the job-level info for each foreign rank?
I mean if some proc is interested in several ranks from the foreign namespace it makes sense to send the job-level data only once.

rhc54 replied:
Yeah, I agree - however, that means tracking who has received the nspace info so we don't send it again. Is it worth that memory footprint? I'm unsure and seeking input.

artpol84 commented on src/server/pmix_server_get.c:
```diff
@@ -404,6 +413,29 @@ static pmix_status_t _satisfy_request(pmix_nspace_t *nptr, int rank,
+        0 != strncmp(nptr->nspace, cd->peer->info->nptr->nspace, PMIX_MAX_NSLEN)) {
+        cur_rank = PMIX_RANK_WILDCARD;
+        if (PMIX_SUCCESS != (rc = pmix_bfrop.pack(&pbkt, &cur_rank, 1, PMIX_INT))) {
+            PMIX_ERROR_LOG(rc);
```
Shouldn't we store the job-level info in the dstor as well?

rhc54 replied:
I think we should, and that would alleviate the problem you flagged above as the client could check and find the job-level info in the dstor (and so we wouldn't have to provide it again). I just wasn't sure how to put the info in there and could use your help.

artpol84 replied:
It's OK to go ahead with this PR as is and we will work on the dstor side later.

artpol84 replied:
I mean in terms of job-info publishing.

rhc54 replied:
Thanks! Appreciate the help

artpol84 commented on src/server/pmix_server_get.c:
```diff
@@ -494,7 +526,7 @@ pmix_status_t pmix_pending_resolve(pmix_nspace_t *nptr, int rank,
                 pmix_status_t rc;
-                rc = _satisfy_request(nptr, rank, req->cbfunc, req->cbdata, NULL);
+                rc = _satisfy_request(nptr, rank, NULL, req->cbfunc, req->cbdata, NULL);
                 if( PMIX_SUCCESS != rc ){
```
@rhc54 why we omitting the server caddy here?
Does this works fine for the following scenario:
1. client from nspace1 asks for rank=1 from nspace2.
2. Server doesn't have this info and issues direct modex request
3. Server receives dmdx response and calls `pmix_pending_resolve()`
it seems that in this case we won't provide the nspace2 info to the client.

rhc54 replied:
Hmmm...let me ponder this one. You well may be correct, but my old brain needs to wrap around it a bit.

rhc54 replied:
Yeah, I agree with your analysis. Will have to find a way to mark the request, or figure it out after dmodex returns

rhc54 replied:
Okay, turned out to be relatively easy as the server_caddy was already available in pending_resolve. Thx!

artpol84 commented on src/server/pmix_server_get.c:
```diff
@@ -404,6 +413,29 @@ static pmix_status_t _satisfy_request(pmix_nspace_t *nptr, int rank,
+         * containing a buffer, so package it accordingly */
+        pbptr = &nptr->server->job_info;
+        if (PMIX_SUCCESS != (rc = pmix_bfrop.pack(&pbkt, &pbptr, 1, PMIX_BUFFER))) {
+            PMIX_ERROR_LOG(rc);
```
What if later we will not find the rank information? we already have this buffer and we will need to just drop it.
I guess that for the performance reasons we want to pack this after we make sure that we have the response now and don't need to go for dmdx.

artpol84 replied:
Also I guess there will be a memory leak in this case:
```C
     PMIX_UNLOAD_BUFFER(&pbkt, data, sz);
     PMIX_DESTRUCT(&pbkt);
 
     if (found) {
         /* pass it back */
         cbfunc(PMIX_SUCCESS, data, sz, cbdata, relfn, data);
         return PMIX_SUCCESS;
     }
```
we assume that `cbfunc` will call `relfn` but if `found` is 0 we won't call `cbfunc` and thus nobody will release `data`

rhc54 replied:
Both points are valid. However, packing a buffer is really cheap and so this may not be a significant issue? If you feel it is, I guess I could put a flag in there to indicate that we need the job info and attach it to the end of the rank info, if found.

I can/will fix the memory leak.

artpol84 replied:
I was about to suggest to just put this after the key search. Will this be ok?

artpol84 replied:
Like this
```C

     if (found) {

PACK NSPACE INFO IF NEED

         /* pass it back */
         cbfunc(PMIX_SUCCESS, data, sz, cbdata, relfn, data);
         return PMIX_SUCCESS;
     }
```

rhc54 replied:
Yeah - no problem, will do! Thx

rhc54 commented:
In addition to the line-based reviews, I've hit something in the OMPI cleanup that bothers me a bit and likely merits more discussion. The issue arose out of support for heterogeneous systems. In this case, each process "puts" its locally detected architecture, the info is circulated as part of the "fence", and then retrieved via "get". However, the architecture can also be provided by the local daemon (so each proc isn't probing it) and even a hetero setup could be known by the RM, thus enabling all arch's to be provided in the job info.

My point here is the issue of what prefix to use when retrieving arch info? If the RM provided it, then it needs to have the "pmix" prefix. However, if the process "puts" it, then we won't find it if we use a key that starts with "pmix" - the process must "put" it with a non-PMIx key.

So I'm wondering if we need to rethink this optimization? I could see adding a check on "put" that returns an error if the key uses a "pmix" prefix - that would at least prevent someone from making the mistake. Alternatively, we could take all "put" data that uses a key with "pmix" prefix and separate it out so it gets stored in the right place on the remote end.

Opinions? I have a slight preference for the latter approach as I don't like returning errors if it can be avoided.

artpol84 commented:
Can PMIx itself probe for architecture? In this case procs will never need to probe it's arch and PMIx is free to use pmix prefix.

rhc54 commented:
I don't think so as this isn't something standardized - every library has their own representation for architecture. It's why the RM has a problem in providing it. In this case, OMPI's internal daemon is "acting" as the RM, and so the values are consistent - but that is the exception case, not the rule.

So we could just define a more generic architecture info that the RM can provide, and then let the libraries convert it to their internal representation. Then the key for this generic value would have a "pmix" prefix, and if the library wants to circulate their internal representation, then that must not have a "pmix" prefix.

Does that make sense?

artpol84 commented:
Sounds OK to me

2016-07-21 23:18 GMT+06:00 rhc54 <notifications@github.com>:

> I don't think so as this isn't something standardized - every library has
> their own representation for architecture. It's why the RM has a problem in
> providing it. In this case, OMPI's internal daemon is "acting" as the RM,
> and so the values are consistent - but that is the exception case, not the
> rule.
>
> So we could just define a more generic architecture info that the RM can
> provide, and then let the libraries convert it to their internal
> representation. Then the key for this generic value would have a "pmix"
> prefix, and if the library wants to circulate their internal
> representation, then that must not have a "pmix" prefix.
>
> Does that make sense?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pmix/master/pull/114#issuecomment-234321790>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHL5PpCUundzvrfKPjkqjge7Y5ke-rPaks5qX6nPgaJpZM4JPEzy>
> .
>



-- 
С Уважением, Поляков Артем Юрьевич
Best regards, Artem Y. Polyakov

rhc54 commented:
I have addressed the comments - if acceptable, I'd like to push off the change regarding the architecture value to another PR, and defer the dstor integration to @artpol84 and friends for after this is committed.

rhc54 commented:
@artpol84 Question arose about whether we really need to make the constraint on rank value - i.e., that all job-related values must be requested with rank "wildcard", and non-job-related values must have their specific rank. This makes things a little harder on users as they have to understand which PMIx keys pertain to job-related data, vs which pertain to proc-specific data.

While the constraint saves us one hash table search, we were a little uncomfortable with imposing that knowledge on the user. It is what tripped us up in OMPI, and I confess I had to think hard on every single pmix put/get call to ensure I had it right.

It occurred to us that the dstor might actually resolve the problem by making the performance hit of looking in two places essentially disappear - we could check the job-info memory location and then check the rank-specific one without any significant time penalty.

Does this sound reasonable to you based on what you guys have done for dstor? Assuming someone asks for a PMIx-prefixed key, can we check job-level data (i.e., data stored with rank wildcard) and then check rank-level data (assuming they gave us a non-wildcard rank) with little penalty?

rhc54 commented:
I suppose we also need to ask: what do we do about backward compatibility here? Do we need to worry about 1.1.4-based apps that didn't follow this rule? In the "cross-version" code, we know what library version the app is using - do we need to take that into account here?

Seems like we probably do...sigh...

ggouaillardet commented:
@rhc54 it seems this has already landed into ompi master.
i found an issue (unpack failure) when ```mpirun -np 1 ./dynamic/spawn```
and the root cause was the nspace was not unpacked.
here is a patch for this branch (from open-mpi/ompi@21e7f31dbe07134e210c5b1eee2a9bb06c752794)

```patch
diff --git a/src/client/pmix_client_get.c b/src/client/pmix_client_get.c
index b167344..714d0e5 100644
--- a/src/client/pmix_client_get.c
+++ b/src/client/pmix_client_get.c
@@ -1,7 +1,7 @@
 /* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
 /*
  * Copyright (c) 2014-2016 Intel, Inc.  All rights reserved.
- * Copyright (c) 2014-2015 Research Organization for Information Science
+ * Copyright (c) 2014-2016 Research Organization for Information Science
  *                         and Technology (RIST). All rights reserved.
  * Copyright (c) 2014      Artem Y. Polyakov <artpol84@gmail.com>.
  *                         All rights reserved.
@@ -298,6 +298,15 @@ static void _getnb_cbfunc(struct pmix_peer_t *pr, pmix_usock_hdr_t *hdr,
         if (PMIX_SUCCESS == (rc = pmix_bfrop.unpack(buf, &bptr, &cnt, PMIX_BUFFER))) {
             /* if the rank is WILDCARD, then this is an nspace blob */
             if (PMIX_RANK_WILDCARD == cur_rank) {
+                char *nspace;
+                /* unpack the nspace - we don't really need it, but have to
+                 * unpack it to maintain sequence */
+                cnt = 1;
+                if (PMIX_SUCCESS != (rc = pmix_bfrop.unpack(bptr, &nspace, &cnt, PMIX_STRING))) {
+                    PMIX_ERROR_LOG(rc);
+                    return;
+                }
+                free(nspace);
                 pmix_client_process_nspace_blob(cb->nspace, bptr);
             } else {
                 cnt = 1;
```

rhc54 commented:
@ggouaillardet Thanks! Yes, this has been on the OMPI master for the last several weeks as some groups needed it there.

rhc54 commented:
@artpol84 Looks like Jenkins is failing to find libevent - can you folks take a look?

artpol84 commented:
@rhc54 ok

artpol84 commented:
bot:retest

artpol84 commented:
looks like libevent migrated from sourceforge to github

rhc54 commented:
oho - interesting!

artpol84 commented:
bot:retest

rhc54 commented:
It occurred to me that this commit isn't complete - for example, if we are going to require that someone specify the nodeid when requesting the number of peers executing on that node, then we must have stored that info as a function of the nodeid. Our current storage code is based solely on _rank_ and has no concept of other indices. With this change, we will need to revamp that system so it can recognize data that is rank-independent and store it accordingly.

rhc54 commented:
Per telecon of Aug 4th:

  * Primary concern was over how to handle node-specific information. Our current storage is based on rank, not nodeid, and so is the PMIx\_Get function. We discussed ways of indicating that the PMIx\_Get call is requesting node-level data, but it rapidly became confusing and clunky.

  * We propose to do the following:
    * Use the PMIx\_Query function to retrieve node-level information, thus avoiding further overload of PMIx\_Get
    * Modify PMIx\_Query to use a new pmix\_query\_t structure that contains:
      * char **keys - identify information being requested
      * pmix\_info\_t qualifiers[] - array of qualifiers to direct the search
      * size\_t nqual - number of qualifiers in array
    * Return function for PMIx\_Query to remain the same, with each pmix\_info\_t in the array to correspond to a pmix\_query\_t. Multiple values returned for a single pmix\_query\_t will be provided in the pmix\_data\_array\_t field in the pmix\_info\_t's value
    * Create a new hash table for node-level data

artpol84 commented:
bot:retest

dsolt commented:
When I use this branch with open-mpi master it fails in MPI_Init.  I'm using the external component for pmix.   I have not tried simply copying the code in place.  Has anyone else tested this with open-mpi master?

rhc54 commented:
It won't work without an update of the OMPI integration as there is further redefinition of the types. I believe I have a branch to support it, but it is also out of date while I work on the node info support.

rhc54 commented:
So go back on vacation, dude! I'll have this up next week when you get back.

rhc54 commented:
I think events have overtaken this PR and it is no longer valid or required
